drivers/usb/uhc/uhc_mcux.c

```diff
@@ -0,0 +1,720 @@
+#include <zephyr/drivers/usb/uhc.h>
+
+#include "uhc_common.h"
+#include "usb.h"
+#include "usb_host_config.h"
+#include "usb_host_mcux_drv_port.h"
+#include "usbh_device.h"
+#ifdef CONFIG_USB_UHC_NXP_PHY
+#include "usb_phy.h"
+#endif
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uhc_mcux, CONFIG_UHC_DRIVER_LOG_LEVEL);
+
+static K_KERNEL_STACK_DEFINE(drv_stack, CONFIG_UHC_MCUX_THREAD_STACK_SIZE);
+
+K_MEM_SLAB_DEFINE_STATIC(mcux_uhc_transfer_pool, sizeof(usb_host_transfer_t),
+			 USB_HOST_CONFIG_MAX_TRANSFERS, sizeof(void *));
+
+#if defined(CONFIG_NOCACHE_MEMORY)
+K_HEAP_DEFINE_NOCACHE(mcux_transfer_alloc_pool, USB_HOST_CONFIG_MAX_TRANSFERS * 8u +
+	1024u * USB_HOST_CONFIG_MAX_TRANSFERS);
+#endif
+
+#define PRV_DATA_HANDLE(_handle) CONTAINER_OF(_handle, struct uhc_mcux_data, host_instance)
+
+#ifdef CONFIG_USB_UHC_NXP_EHCI
+#include "usb_host_ehci.h"
+static const usb_host_controller_interface_t mcux_ehci_usb_iface = {
+	USB_HostEhciCreate,	USB_HostEhciDestory,  USB_HostEhciOpenPipe, USB_HostEhciClosePipe,
+	USB_HostEhciWritePipe, USB_HostEhciReadpipe, USB_HostEhciIoctl,
+};
+#endif
+
+#ifdef CONFIG_USB_UHC_NXP_KHCI
+#include "usb_host_khci.h"
+static const usb_host_controller_interface_t mcux_khci_usb_iface = {
+	USB_HostKhciCreate,	USB_HostKhciDestory,  USB_HostKhciOpenPipe, USB_HostKhciClosePipe,
+	USB_HostKhciWritePipe, USB_HostKhciReadpipe, USB_HostKciIoctl,
+};
+#endif
+
+struct uhc_mcux_ep_handle {
+	uint8_t ep;
+	void *udev;
+	usb_host_pipe_handle mcux_ep_handle;
+};
+
+struct uhc_mcux_data {
+	const struct device *dev;
+	struct uhc_mcux_ep_handle ep_handles[USB_HOST_CONFIG_MAX_PIPES];
+	usb_host_instance_t host_instance;
+	struct k_thread drv_stack_data;
+	bool need_update_addr;
+	uint8_t top_device_speed;
+	uint8_t controller_id;
+};
+
+struct uhc_mcux_config {
+	const usb_host_controller_interface_t *mcux_if;
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+	uint32_t base;
+#ifdef CONFIG_USB_UHC_NXP_PHY
+	uint8_t need_to_init_phy;
+	uint8_t D_CAL;
+	uint8_t TXCAL45DP;
+	uint8_t TXCAL45DM;
+#endif
+};
+
+#if defined(CONFIG_NOCACHE_MEMORY)
+static void *uhc_mcux_nocache_alloc(uint32_t size)
+{
+	void *p = (void *)k_heap_alloc(&mcux_transfer_alloc_pool, size, K_NO_WAIT);
+
+	if (p != NULL) {
+		(void)memset(p, 0, size);
+	}
+
+	return p;
+}
+
+static void uhc_mcux_nocache_free(void *p)
+{
+	k_heap_free(&mcux_transfer_alloc_pool, p);
+}
+#endif
+
+static int uhc_mcux_bus_control(const struct device *dev, usb_host_bus_control_t type)
+{
+	const struct uhc_mcux_config *config = dev->config;
+	struct uhc_data *data = dev->data;
+	struct uhc_mcux_data *priv = data->priv;
+	usb_status_t status;
+
+	status = config->mcux_if->controllerIoctl(
+			priv->host_instance.controllerHandle, kUSB_HostBusControl, &type);
+	if (status != kStatus_USB_Success) {
+		return -EIO;
+	}
+	return 0;
+}
+
+usb_status_t USB_HostAttachDevice(usb_host_handle hostHandle,
```

Comment:
I believe I made this comment earlier but I wanted to suggest it again- could we add comments above each of the `USB_Hostxxx` functions, like the following:

```
/*
 * Note- this function will be called by the MCUX SDK host controller layer when...
 */
```

For example:
```
/*
 * Note- this function will be called by the MCUX SDK host controller layer when a USB device is attached
 */
```

I think this will help future developers working with the driver understand the use of these functions, since the fact that they are callbacks is not immediately obvious.